content/ADR-214-pointer-events-sdk-component.md

```diff
@@ -0,0 +1,197 @@
+PointerEventResults
+
+ROOT_ENTITY
+  └── A ✅ (PointerEvents, MeshCollider)
+```
+
+```
+In this scenario, the entity B has a MeshCollider, but since the entity
+does not have one, it will not intercept any event.
+
+ROOT_ENTITY
+  └── A ❌ (PointerEvents)
+      └── B ❌ (MeshCollider)
+```
+
+```
+In this scenario, the entity A has a GltfContainer and PointerEvents.
+Like in the SDK entities, the interal collider meshes of the
+GltfContainer will be used to filter which meshes are eligible for raycasts.
+
+ROOT_ENTITY
+  └── A ❌ (GltfContainner, PointerEvents) <─┐
+      ├── GLTF_ROOT                          │
+      │   └── ...                            │
+      │       └── ✅ internal gltf node (MeshCollider)
+      └── B ❌ (MeshCollider)
+```
+
+### Mesh selection for UiElements
+
+Any entity with `UiTransform` ([ADR-124](/adr/ADR-124)) can be eligible for pointer events if they have a `PointerEvents` component, all other UI entities will not be eligible and pointer events will be bubbled up until finding an entity that matches the criteria. A rectangular mesh occupying the whole area designated by the `UiTransform` MUST be used.
+
+In UI entities, contrary to the render order ([ADR-151](/adr/ADR-151)), the events MUST bubble up from the leaves of the tree up to the root until finding an entity that matches the pointer event. In that case, the propagation of the event MUST stop, and the event MUST be added to the `PointerEventsResult`.
+
+## Input commands
+
+For the ECS to work, "events" and "state" must be separated into two different categories. A challenge is faced here, since "events" or "interrupts" are not compatible with data representations at one moment in time (like the state of the ECS). To overcome this, all the input events of the frame must be queued and processed at the "executeRaycast" stage of the tick for all the scenes that are running. Naturally, this compute MUST count towards the quota/limit suggested by ADR-148.
+
+Then, in the "executeRaycast" step of the scene tick (on the renderer side), that queue will be consumed and "InputCommands" will be produced. An InputCommand is a record in a PointerEventResult component listing all the pertinent input actions for that entity. It is done to process when the pointer is "hovering" an entity, when it leaves, when the key down and key up in input. All of that can happen in the same frame. For example, a cube can be aimed at, clicked, and released the click on the same frame.
+
+This design enables rapid games with low input lag and without losing information regardless of the framerate of the scene and renderer.
+
+It also enables high resolution events for cases of moving entities, e.g. assuming one is pointing still, with one finger on the trigger, if the engine puts an entity in front of someone for one frame, the scene will receive the "Hover Enter" event for that unique frame, and if one is fast enough, it may also be clicked.
```

Comment:
```suggestion
It also enables high resolution events for cases of moving entities, e.g. assuming one is pointing still, with one finger on the trigger, if the engine puts an entity in front of someone for one frame, the scene will receive the "HoverEnter" event for that unique frame, and if one is fast enough, it may also be clicked.
```